From a124abd6b9944d24a4327afc78f2b048fae3bd50 Mon Sep 17 00:00:00 2001
From: Torge Matthies <openglfreak@googlemail.com>
Date: Fri, 14 Jun 2024 03:11:44 +0200
Subject: [PATCH 29/42] winegstreamer: Handle Gstreamer pipeline flushes
 gracefully.

If a GST_EVENT_FLUSH_START event is received between the calls from PE code to wg_parser_stream_get_buffer
and wg_parser_stream_copy_buffer, the latter function will return an error, resulting in the sample request being
dropped without having delivered a sample.

If wg_parser_stream_copy_buffer returns an error, retry the whole wait_on_sample procedure.
---
 dlls/winegstreamer/media_source.c | 12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

diff --git a/dlls/winegstreamer/media_source.c b/dlls/winegstreamer/media_source.c
index 58b82a444b4..7808d3966ff 100644
--- a/dlls/winegstreamer/media_source.c
+++ b/dlls/winegstreamer/media_source.c
@@ -635,6 +635,7 @@ static HRESULT media_stream_send_sample(struct media_stream *stream, const struc
 
     if (!wg_parser_stream_copy_buffer(stream->wg_stream, data, 0, wg_buffer->size))
     {
+        hr = HRESULT_FROM_WIN32(ERROR_RETRY);
         wg_parser_stream_release_buffer(stream->wg_stream);
         IMFMediaBuffer_Unlock(buffer);
         goto out;
@@ -693,10 +694,10 @@ static HRESULT wait_on_sample(struct media_stream *stream, IUnknown *token)
 {
     struct media_source *source = impl_from_IMFMediaSource(stream->media_source);
     struct wg_parser_buffer buffer;
+    HRESULT hr;
     BOOL ret;
 
-    TRACE("%p, %p\n", stream, token);
-
+retry:
     stream->busy = TRUE;
     LeaveCriticalSection(&source->cs);
     memset(&buffer, 0, sizeof(buffer));
@@ -712,7 +713,12 @@ static HRESULT wait_on_sample(struct media_stream *stream, IUnknown *token)
     }
 
     if (ret)
-        return media_stream_send_sample(stream, &buffer, token);
+    {
+        hr = media_stream_send_sample(stream, &buffer, token);
+        if (hr == HRESULT_FROM_WIN32(ERROR_RETRY))
+            goto retry;
+        return hr;
+    }
     return media_stream_send_eos(source, stream);
 }
 
-- 
2.45.2

